File: Leetcode-one-by-one/src/Special Binary String.swift
class Solution {
    func makeLargestSpecial(_ s: String) -> String {
        let chars = Array(s)
        return helper(chars, 0, chars.count)
    }

    private func helper(_ chars: [Character], _ start: Int, _ end: Int) -> String {
        var balance = 0
        var last = start
        var subs: [String] = []

        for i in start..<end {
            if chars[i] == "1" {
                balance += 1
            } else {
                balance -= 1
            }

            // Found a primitive special substring
            if balance == 0 {
                // recurse on inner part
                let inner = helper(chars, last + 1, i)
                subs.append("1" + inner + "0")
                last = i + 1
            }
        }

        // sort descending for lexicographically largest result
        subs.sort(by: >)
        return subs.joined()
    }
}

File: Leetcode-one-by-one/src/Special Binary String.java
import java.util.*;

class Solution {
    public String makeLargestSpecial(String s) {
        List<String> parts = new ArrayList<>();
        int balance = 0;
        int last = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '1') balance++;
            else balance--;

            // Found a primitive special substring
            if (balance == 0) {
                // Recursively process the inner substring
                String inner = makeLargestSpecial(s.substring(last + 1, i));
                parts.add("1" + inner + "0");
                last = i + 1;
            }
        }

        // Sort descending for lexicographically largest result
        Collections.sort(parts, Collections.reverseOrder());

        // Concatenate
        StringBuilder sb = new StringBuilder();
        for (String part : parts) {
            sb.append(part);
        }

        return sb.toString();
    }
}
